MultipleInstantiationsOfSingletons.hasSynchronized() can produce stack overflow

When running latest spotbugs version 4.9.0 against my code I get the Exception below. I did not have time yet to find out which exact code in my code base is triggering this so I cannot give a reproducer yet.

```
`Exception in thread "main" edu.umd.cs.findbugs.ba.AnalysisException: Exception was thrown during analysis
	at edu.umd.cs.findbugs.FindBugs2.lambda$analyzeApplication$3(FindBugs2.java:1117)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at edu.umd.cs.findbugs.FindBugs2.analyzeApplication(FindBugs2.java:1110)
	at edu.umd.cs.findbugs.FindBugs2.execute(FindBugs2.java:309)
	at edu.umd.cs.findbugs.FindBugs.runMain(FindBugs.java:399)
	at edu.umd.cs.findbugs.FindBugs2.main(FindBugs2.java:1223)
Caused by: java.util.concurrent.ExecutionException: java.lang.StackOverflowError
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at edu.umd.cs.findbugs.FindBugs2.lambda$analyzeApplication$3(FindBugs2.java:1112)
	... 5 more
Caused by: java.lang.StackOverflowError
	at java.base/java.util.ArrayList.indexOf(ArrayList.java:318)
	at java.base/java.util.ArrayList.contains(ArrayList.java:307)
	at edu.umd.cs.findbugs.detect.MultipleInstantiationsOfSingletons.hasSynchronized(MultipleInstantiationsOfSingletons.java:280)
	at edu.umd.cs.findbugs.detect.MultipleInstantiationsOfSingletons.hasSynchronized(MultipleInstantiationsOfSingletons.java:287)
	at edu.umd.cs.findbugs.detect.MultipleInstantiationsOfSingletons.hasSynchronized(MultipleInstantiationsOfSingletons.java:287)
	at edu.umd.cs.findbugs.detect.MultipleInstantiationsOfSingletons.hasSynchronized(MultipleInstantiationsOfSingletons.java:287)
	at edu.umd.cs.findbugs.detect.MultipleInstantiationsOfSingletons.hasSynchronized(MultipleInstantiationsOfSingletons.java:287)
	at edu.umd.cs.findbugs.detect.MultipleInstantiationsOfSingletons.hasSynchronized(MultipleInstantiationsOfSingletons.java:287)
	at edu.umd.cs.findbugs.detect.MultipleInstantiationsOfSingletons.hasSynchronized(MultipleInstantiationsOfSingletons.java:287)
	at edu.umd.cs.findbugs.detect.MultipleInstantiationsOfSingletons.hasSynchronized(MultipleInstantiationsOfSingletons.java:287)
	at edu.umd.cs.findbugs.detect.MultipleInstantiationsOfSingletons.hasSynchronized(MultipleInstantiationsOfSingletons.java:287)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ Currently the versioning policy of this project follows [Semantic Versioning v2.
 - Accept the 1st parameter of `java.util.concurrent.CompletableFuture`'s `completeOnTimeout()`, `getNow()` and `obtrudeValue()` functions as nullable ([#1001](https://github.com/spotbugs/spotbugs/issues/1001)).
 - Fixed the an analysis error when `FindReturnRef` was checking instructions corresponding to a CFG branch that was optimized away ([#3266](https://github.com/spotbugs/spotbugs/issues/3266))
 - Added execute file permission to files in the distribution archive ([#3274](https://github.com/spotbugs/spotbugs/issues/3274))
+- Fixed a stack overflow in  `MultipleInstantiationsOfSingletons` when a singleton initializer makes recursive calls ([#3280](https://github.com/spotbugs/spotbugs/issues/3280))
 
 ## 4.9.0 - 2025-01-15
 ### Added

--- a/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/Issue3280Test.java
+++ b/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/Issue3280Test.java
@@ -10,5 +10,6 @@ class Issue3280Test extends AbstractIntegrationTest {
         performAnalysis("ghIssues/Issue3280.class");
 
         assertNoBugType("SING_SINGLETON_IMPLEMENTS_CLONEABLE");
+        assertBugInClass("SING_SINGLETON_GETTER_NOT_SYNCHRONIZED", "Issue3280");
     }
 }

--- a/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/Issue3280Test.java
+++ b/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/Issue3280Test.java
@@ -1,0 +1,14 @@
+package edu.umd.cs.findbugs.detect;
+
+import edu.umd.cs.findbugs.AbstractIntegrationTest;
+import org.junit.jupiter.api.Test;
+
+class Issue3280Test extends AbstractIntegrationTest {
+
+    @Test
+    void testIssue() {
+        performAnalysis("ghIssues/Issue3280.class");
+
+        assertNoBugType("SING_SINGLETON_IMPLEMENTS_CLONEABLE");
+    }
+}

--- a/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/Issue3280Test.java
+++ b/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/Issue3280Test.java
@@ -9,7 +9,7 @@ class Issue3280Test extends AbstractIntegrationTest {
     void testIssue() {
         performAnalysis("ghIssues/Issue3280.class");
 
-        assertNoBugType("SING_SINGLETON_IMPLEMENTS_CLONEABLE");
-        assertBugInClass("SING_SINGLETON_GETTER_NOT_SYNCHRONIZED", "Issue3280");
+        assertNoBugInClass("SING_SINGLETON_IMPLEMENTS_CLONEABLE", "Issue3280");
+        assertBugInClassCount("SING_SINGLETON_GETTER_NOT_SYNCHRONIZED", "Issue3280", 1);
     }
 }

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MultipleInstantiationsOfSingletons.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MultipleInstantiationsOfSingletons.java
@@ -243,7 +243,7 @@ public class MultipleInstantiationsOfSingletons extends OpcodeStackDetector {
             }
         }
 
-        if (instanceGetterMethod != null && !hasSynchronized(instanceGetterMethod) && isInstanceFieldLazilyInitialized) {
+        if (instanceGetterMethod != null && !hasSynchronized(instanceGetterMethod, new HashSet<>()) && isInstanceFieldLazilyInitialized) {
             bugReporter.reportBug(new BugInstance(this, "SING_SINGLETON_GETTER_NOT_SYNCHRONIZED", NORMAL_PRIORITY).addClass(this)
                     .addMethod(instanceGetterMethod));
         }
@@ -276,7 +276,17 @@ public class MultipleInstantiationsOfSingletons extends OpcodeStackDetector {
         super.visitAfter(javaClass);
     }
 
-    private boolean hasSynchronized(XMethod method) {
+    /**
+     * @param method The method we are checking
+     * @param visitedMethods The set of methods we have already visited, to break recursion when methods are called recursively. To be initially empty.
+     * @return true if the method (or one of the methods called recursively) is synchronized or uses a monitor.
+     */
+    private boolean hasSynchronized(XMethod method, Set<XMethod> visitedMethods) {
+        if (!visitedMethods.add(method)) {
+            // We are visiting an already visited method and have not found synchronization so far
+            return false;
+        }
+
         if (method.isSynchronized() || methodsUsingMonitor.contains(method)) {
             return true;
         }
@@ -284,7 +294,7 @@ public class MultipleInstantiationsOfSingletons extends OpcodeStackDetector {
         if (calledMethodsByMethods.containsKey(method)) {
             List<XMethod> calledMethods = calledMethodsByMethods.get(method);
             for (XMethod cm : calledMethods) {
-                if (hasSynchronized(cm)) {
+                if (hasSynchronized(cm, visitedMethods)) {
                     return true;
                 }
             }

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MultipleInstantiationsOfSingletons.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MultipleInstantiationsOfSingletons.java
@@ -173,23 +173,24 @@ public class MultipleInstantiationsOfSingletons extends OpcodeStackDetector {
             }
         } else if (seen == Const.ATHROW && "clone".equals(getMethodName()) && stack.getStackDepth() > 0) {
             OpcodeStack.Item item = stack.getStackItem(0);
-            if (item != null && "Ljava/lang/CloneNotSupportedException;".equals(item.getSignature()) && cloneOnlyThrowsException) {
+            if (item != null && cloneOnlyThrowsException && "Ljava/lang/CloneNotSupportedException;".equals(item.getSignature())) {
                 cloneOnlyThrowsCloneNotSupportedException = true;
             }
         } else if (seen == Const.ARETURN && stack.getStackDepth() > 0) {
             OpcodeStack.Item item = stack.getStackItem(0);
             XMethod method = getXMethod();
             XField field = item.getXField();
-            if (isInstanceField(field, getClassName()) && method.isPublic() && method.isStatic()) {
+            if (method.isPublic() && method.isStatic() && isInstanceField(field, getClassName())) {
                 instanceGetterMethods.put(field, method);
             } else {
                 XMethod calledMethod = item.getReturnValueOf();
                 SignatureParser parser = new SignatureParser(getMethodSig());
                 String calledMethodReturnType = ClassName.fromFieldSignature(parser.getReturnTypeSignature());
 
-                if (calledMethod != null && Const.CONSTRUCTOR_NAME.equals(calledMethod.getName())
+                if (calledMethod != null && !method.isPrivate() && method.isStatic()
+                        && Const.CONSTRUCTOR_NAME.equals(calledMethod.getName())
                         && calledMethod.getClassName().equals(getDottedClassName())
-                        && !method.isPrivate() && method.isStatic() && getClassName().equals(calledMethodReturnType)
+                        && getClassName().equals(calledMethodReturnType)
                         && !Const.CONSTRUCTOR_NAME.equals(getMethodName()) && !Const.STATIC_INITIALIZER_NAME.equals(getMethodName())
                         && !("clone".equals(getMethodName()) && "()Ljava/lang/Object;".equals(getMethodSig()))) {
                     hasNoFactoryMethod = false;
@@ -225,11 +226,9 @@ public class MultipleInstantiationsOfSingletons extends OpcodeStackDetector {
         // - the instance field is either eagerly or lazily initialized
 
         if (!(hasSingletonPostFix
-                || (!javaClass.isAbstract() && !javaClass.isInterface() && !javaClass.isRecord()
-                        && isInstanceAssignOk
-                        && hasNoFactoryMethod
-                        && instanceGetterMethod != null
-                        && (isInstanceFieldEagerlyInitialized || isInstanceFieldLazilyInitialized)))) {
+                || (isInstanceAssignOk && hasNoFactoryMethod && instanceGetterMethod != null
+                        && (isInstanceFieldEagerlyInitialized || isInstanceFieldLazilyInitialized)
+                        && !javaClass.isAbstract() && !javaClass.isInterface() && !javaClass.isRecord()))) {
             return;
         }
 
@@ -243,7 +242,7 @@ public class MultipleInstantiationsOfSingletons extends OpcodeStackDetector {
             }
         }
 
-        if (instanceGetterMethod != null && !hasSynchronized(instanceGetterMethod, new HashSet<>()) && isInstanceFieldLazilyInitialized) {
+        if (instanceGetterMethod != null && isInstanceFieldLazilyInitialized && !hasSynchronized(instanceGetterMethod, new HashSet<>())) {
             bugReporter.reportBug(new BugInstance(this, "SING_SINGLETON_GETTER_NOT_SYNCHRONIZED", NORMAL_PRIORITY).addClass(this)
                     .addMethod(instanceGetterMethod));
         }

--- a/spotbugsTestCases/src/java/ghIssues/Issue3280.java
+++ b/spotbugsTestCases/src/java/ghIssues/Issue3280.java
@@ -1,0 +1,23 @@
+package ghIssues;
+
+public class Issue3280 {
+    private static volatile Issue3280 instance = null;
+
+    public static Issue3280 getInstance() {
+        if (null == instance) {
+            instance = new Issue3280();
+        }
+        instance.factorial(5);
+        return instance;
+    }
+
+    private Issue3280() {
+    }
+
+    private int factorial(int n) {
+        if (n == 0) {
+            return 1;
+        }
+        return n * factorial(n - 1);
+    }
+}

